### PR TITLE
Fixes 2940: add snapshot rpm list api

### PR DESCRIPTION
**Pull request by jlsherrill**
## Summary
adds a new snapshot rpm list api:

```// @Router       /repositories/{uuid}/rpms [get]```

## Testing steps

Create a repo and let it snapshot, then fetch the rpms:

```
GET http://localhost:8000/api/content-sources/v1.0/snapshots/0e6c9e7f-e021-4710-8021-f59cb9c56d82/rpms?search=foo&limit=1
```

## Checklist

- [ ] Tested with snapshotting feature disabled and pulp server URL not configured if appropriate

### CODE REVIEW

**Comment by jlsherrill** [comment]
requires https://github.com/content-services/tang/pull/5

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-2940

**Comment by swadeley** [comment]
@jlsherrill Please rebase

**Review by rverdile — COMMENTED** [review]
I was testing this alongside the Tang PR. Should be good once it's rebased.

File: pkg/api/rpms.go [hunk 1]
```diff
@@ -11,12 +11,27 @@ type RepositoryRpm struct {
 	Checksum string `json:"checksum"` // The checksum of the rpm
 }
 
+type SnapshotRpm struct {
+	Name    string `json:"name"`    // The rpm package name
+	Arch    string `json:"arch"`    // The Architecture of the rpm
```
Comment (rverdile) [inline pkg/api/rpms.go:16]:
```suggestion
	Arch    string `json:"arch"`    // The architecture of the rpm
```

File: pkg/dao/rpms_test.go [hunk 2]
```diff
@@ -953,3 +953,48 @@ func (s *RpmSuite) TestSearchRpmsForSnapshots() {
...
+	repoConfig := models.RepositoryConfiguration{}
+	res := s.tx.Where("org_id = ?", orgId).First(&repoConfig)
+	require.NoError(s.T(), res.Error)
+	snaps, err := seeds.SeedSnapshots(s.tx, repoConfig.UUID, 1)
+	require.NoError(s.T(), err)
+	res = s.tx.Model(models.Snapshot{}).Where("repository_configuration_uuid = ?", repoConfig.UUID).Update("version_href", hrefs[0])
+	require.NoError(s.T(), res.Error)
+
+	// pkgs, total, err := (*config.Tang).RpmRepositoryVersionPackageList(ctx, pulpHrefs, tangy.RpmListFilters{Name: search}, tangy.PageOptions{
+
```
Comment (rverdile) [inline pkg/dao/rpms_test.go:984]: commented out function call hiding here

**Comment by jlsherrill** [comment]
updated, rebased, and set to use the new tang

**Review by rverdile — APPROVED** [review]
nice!

**Comment by swadeley** [comment]
Hi

`In [4]: app.content_sources.rest_client.rpms_api.list_repositories_rpms.params_map
Out[4]: 
{'all': ['uuid',
  'limit',
  'offset',
  'search',
  'sort_by',
  'async_req',
  '_host_index',
  '_preload_content',
  '_request_timeout',
  '_return_http_data_only',
  '_check_input_type',
  '_check_return_type'],
 'required': ['uuid'],
 'nullable': ['_request_timeout'],
 'enum': [],
 'validation': []}

In [5]: created_repo['uuid']
Out[5]: '45c35a10-420c-4507-92c1-b762b5635a6f'

In [6]: app.content_sources.rest_client.rpms_api.list_repositories_rpms('45c35a10-420c-4507-92c1-b762b5635a6f')
2024-03-08 19:01:26.677 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<>, params=[]
Out[6]: 
{'data': [{'arch': 'noarch',
           'checksum': '7a831f9f90bf4d21027572cb503d20b702de8e8785b02c0397445c2e481d81b3',
           'epoch': 0,
           'name': 'bear',
           'release': '1',
           'summary': 'A dummy package of bear',
           'uuid': 'ba75ff00-4785-4347-82a8-ef17dde8397a',
           'version': '4.1'},
`

```
In [9]: app.content_sources.rest_client.rpms_api.list_repositories_rpms('45c35a10-420c-4507-92c1-b762b5635a6f', limit=2, offset=2)
2024-03-08 19:03:36.129 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<>, params=[('limit', 2), ('offset', 2)]
Out[9]: 
{'data': [{'arch': 'noarch',
           'checksum': '43e77adb7f51b5542b813024a8ee3e477175c142f35982ab5ae2b2978486239f',
           'epoch': 0,
           'name': 'cat',
           'release': '1',
           'summary': 'A dummy package of cat',
           'uuid': '57d987f6-b369-4076-90aa-1000aa0bb924',
           'version': '1.0'},
          {'arch': 'noarch',
           'checksum': '2189ac4bf059f98a8d48136eb72b46415f3aa263026443ebd8879d415eacaf42',
           'epoch': 0,
           'name': 'cheetah',
           'release': '5',
           'summary': 'A dummy package of cheetah',
           'uuid': 'eb869ce3-a17f-4625-8442-46ee0ba696d0',
           'version': '1.25.3'}],
 'links': {'first': '/api/content-sources/v1/repositories/45c35a10-420c-4507-92c1-b762b5635a6f/rpms?limit=2&offset=0',
           'last': '/api/content-sources/v1/repositories/45c35a10-420c-4507-92c1-b762b5635a6f/rpms?limit=2&offset=30',
           'next': '/api/content-sources/v1/repositories/45c35a10-420c-4507-92c1-b762b5635a6f/rpms?limit=2&offset=4',
           'prev': '/api/content-sources/v1/repositories/45c35a10-420c-4507-92c1-b762b5635a6f/rpms?limit=2&offset=0'},
 'meta': {'count': 32, 'limit': 2, 'offset': 2}}

In [10]: 
```

lgtm

**Comment by swadeley** [comment]
Hows this:

```In [19]: app.content_sources.rest_client.rpms_api.search_snapshot_rpms(dict(uuids=['4f919c63-e47f-470a-9578-8828a89eecda']))
2024-03-08 19:32:16.327 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=<>, params=[]
Out[19]: 
[{'package_name': 'bear', 'summary': 'A dummy package of bear'},
 {'package_name': 'camel', 'summary': 'A dummy package of camel'},
 {'package_name': 'cat', 'summary': 'A dummy package of cat'},
 {'package_name': 'cheetah', 'summary': 'A dummy package of cheetah'},
 {'package_name': 'chimpanzee', 'summary': 'A dummy package of chimpanzee'},
 {'package_name': 'cockateel', 'summary': 'A dummy package of cockateel'},
 {'package_name': 'cow', 'summary': 'A dummy package of cow'},
 {'package_name': 'crow', 'summary': 'A dummy package of crow'},
 {'package_name': 'dog', 'summary': 'A dummy package of dog'},
 {'package_name': 'dolphin', 'summary': 'A dummy package of dolphin'},
 {'package_name': 'duck', 'summary': 'A dummy package of duck'},
 {'package_name': 'elephant', 'summary': 'A dummy package of elephant'},
 {'package_name': 'fox', 'summary': 'A dummy package of fox'},
 {'package_name': 'frog', 'summary': 'A dummy package of frog'},
 {'package_name': 'giraffe', 'summary': 'A dummy package of giraffe'},
 {'package_name': 'gorilla', 'summary': 'A dummy package of gorilla'},
 {'package_name': 'horse', 'summary': 'A dummy package of horse'},
 {'package_name': 'kangaroo', 'summary': 'A dummy package of kangaroo'},
 {'package_name': 'lion', 'summary': 'A dummy package of lion'},
 {'package_name': 'mouse', 'summary': 'A dummy package of mouse'},
 {'package_name': 'penguin', 'summary': 'A dummy package of penguin'},
 {'package_name': 'pike', 'summary': 'A dummy package of pike'},
 {'package_name': 'shark', 'summary': 'A dummy package of shark'},
 {'package_name': 'squirrel', 'summary': 'A dummy package of squirrel'},
 {'package_name': 'stork', 'summary': 'A dummy package of stork'},
 {'package_name': 'tiger', 'summary': 'A dummy package of tiger'},
 {'package_name': 'trout', 'summary': 'A dummy package of trout'},
 {'package_name': 'walrus', 'summary': 'A dummy package of walrus'},
 {'package_name': 'whale', 'summary': 'A dummy package of whale'},
 {'package_name': 'wolf', 'summary': 'A dummy package of wolf'},
 {'package_name': 'zebra', 'summary': 'A dummy package of zebra'}]

```

**Comment by swadeley** [comment]
Hi, in master branch I already list rpms and search for one in the list for a repo in stage:
```

In [8]: app.content_sources.rest_client.rpms_api.list_repositories_rpms('07a581c4-e586-4eee-b853-042f77301bdc', search='acme-package')
2024-03-11 09:20:11.941 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=892fcf8fad22485cb62b4d217953ad80, 892fcf8fad22485cb62b4d217953ad80, params=[('search', 'acme-package')]
Out[8]: 
{'data': [{'arch': 'noarch',
           'checksum': '2f7db1e01ccb0e0c0c4a6ef4981e7398927e5e558671c5989c6b667a8c702e56',
           'epoch': 0,
           'name': 'acme-package',
           'release': '1.elfake',
           'summary': 'acme-package package',
           'uuid': '258a5f0a-10f6-40f8-8350-188218ee08d5',
           'version': '1.0.2'},
          {'arch': 'noarch',
           'checksum': 'da1bdb74dc9687adfa7e1a85026c2373ed8d91c03639446d9b865c3bf3e9c67d',
           'epoch': 0,
           'name': 'acme-package',
           'release': '1.elfake',
           'summary': 'acme-package package',
           'uuid': '603d140d-5093-4f89-a61c-ad9d5605c67a',
           'version': '1.1.2'},
          {'arch': 'noarch',
           'checksum': 'e5fae88537edadbda567359686451986b7947fa26b7a0bdc68fbf0373dcecbd1',
           'epoch': 0,
           'name': 'acme-package',
           'release': '1.elfake',
           'summary': 'acme-package package',
           'uuid': 'de386858-310d-4611-b13c-179977319cb0',
           'version': '1.0.1'}],
 'links': {'first': '/api/content-sources/v1/repositories/07a581c4-e586-4eee-b853-042f77301bdc/rpms?limit=100&offset=0&search=acme-package',
           'last': '/api/content-sources/v1/repositories/07a581c4-e586-4eee-b853-042f77301bdc/rpms?limit=100&offset=0&search=acme-package'},
 'meta': {'count': 3, 'limit': 100, 'offset': 0}}
```

**Comment by swadeley** [comment]
Hi, on master branch, I reinstalled the plugin, opened shell to stage, created repo, got snapshot UUID:

`   'last_snapshot_uuid': 'c2eedc6f-d57f-46bd-a933-89b351de6f3`

Ran this:
In [2]: app.content_sources.rest_client.rpms_api.search_snapshot_rpms(dict(uuids=['c2eedc6f-d57f-46bd-a933-89b351de6f31']))

**ServiceException: (500)
Reason: Internal Server Error**

So search_snapshot_rpms is the new path, **search** not **list**. OK?

**Comment by swadeley** [comment]
More with search and limit.

I used this repo:
`https://mmccune.fedorapeople.org/repos/small/`
which has three acme packages but the snapshot only has one, so testing with **limit** is limited in this case:
```

In [5]: app.content_sources.rest_client.rpms_api.search_snapshot_rpms(dict(uuids=['417eaa78-d12c-488e-a1f8-12f5dc359168']))
2024-03-11 09:54:54.122 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=<>, params=[]
Out[5]: 
[{'package_name': 'acme-package', 'summary': 'acme-package package'},
 {'package_name': 'beautifulness', 'summary': 'beautifulness package'},
 {'package_name': 'diphtheric', 'summary': 'diphtheric package'},
 {'package_name': 'frugalist', 'summary': 'frugalist package'},
 {'package_name': 'maleability', 'summary': 'maleability package'}]

In [6]: app.content_sources.rest_client.rpms_api.search_snapshot_rpms(dict(uuids=['417eaa78-d12c-488e-a1f8-12f5dc359168'], search='frugalist'))
2024-03-11 09:56:41.442 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=<>, params=[]
Out[6]: [{'package_name': 'frugalist', 'summary': 'frugalist package'}]

In [7]: app.content_sources.rest_client.rpms_api.search_snapshot_rpms(dict(uuids=['417eaa78-d12c-488e-a1f8-12f5dc359168'], search='acme-package'))
2024-03-11 09:59:26.052 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=<>, params=[]
Out[7]: [{'package_name': 'acme-package', 'summary': 'acme-package package'}]


In [9]: app.content_sources.rest_client.rpms_api.search_snapshot_rpms(dict(uuids=['417eaa78-d12c-488e-a1f8-12f5dc359168'], search='acme-package', limit=1))
2024-03-11 10:01:17.090 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=<>, params=[]
Out[9]: [{'package_name': 'acme-package', 'summary': 'acme-package package'}]
```

**Comment by jlsherrill** [comment]
search is an existing api for searching for rpm names, this api is 'list', i'd expect it to be called list_snapshot_rpms or something similar

**Comment by jlsherrill** [comment]
I didn't realize it would be duplicated into each 'tag', so i've updated here to only use the snapshot tag.  IDK if that will make a difference, but it really doesn't need to be in 3 places in the generated client.  Maybe try regenerating and see if it shows up in the snapshot_api?

File: pkg/handler/rpms.go [hunk 3]
```diff
@@ -130,3 +131,39 @@ func (rh *RpmHandler) searchSnapshotRPMs(c echo.Context) error {
...
+// @ID           listSnapshotRpms
+// @Description  List RPMs in a repository snapshot.
+// @Tags         snapshots
+// @Accept       json
+// @Produce      json
+// @Param		 uuid	path string true "Snapshot ID."
+// @Param		 limit query int false "Number of items to include in response. Use it to control the number of items, particularly when dealing with large datasets. Default value: `100`."
+// @Param		 offset query int false "Starting point for retrieving a subset of results. Determines how many items to skip from the beginning of the result set. Default value:`0`."
+// @Param		 search query string false "Term to filter and retrieve items that match the specified search criteria. Search term can include name."
+// @Success      200 {object} api.RepositoryRpmCollectionResponse
```
Comment (rverdile) [inline pkg/handler/rpms.go:146]: I think it's because this should be "SnapshotRpmCollectionResponse" ?

Comment (jlsherrill) [reply pkg/handler/rpms.go]: thanks! updated

**Comment by swadeley** [comment]
Hi, all working now:


`In [2]: app.content_sources.rest_client.snapshots_api.list_snapshot_rpms('b342c4f3-c0bd-4eb4-9b2d-aad494928d2d', limit=4)
2024-03-13 18:34:30.499 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<>, params=[('limit', 4)]
Out[2]: 
{'data': [{'arch': 'noarch',
           'epoch': '0',
           'name': 'bear',
           'release': '1',
           'summary': 'A dummy package of bear',
           'version': '4.1'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'camel',
           'release': '1',
           'summary': 'A dummy package of camel',
           'version': '0.1'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'cat',
           'release': '1',
           'summary': 'A dummy package of cat',
           'version': '1.0'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'cheetah',
           'release': '5',
           'summary': 'A dummy package of cheetah',
           'version': '1.25.3'}],
 'links': {'first': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=4&offset=0',
           'last': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=4&offset=32',
           'next': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=4&offset=4'},
 'meta': {'count': 35, 'limit': 4, 'offset': 0}}

In [3]: app.content_sources.rest_client.snapshots_api.list_snapshot_rpms('b342c4f3-c0bd-4eb4-9b2d-aad494928d2d', search='cat')
2024-03-13 18:34:50.674 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<>, params=[('search', 'cat')]
Out[3]: 
{'data': [{'arch': 'noarch',
           'epoch': '0',
           'name': 'cat',
           'release': '1',
           'summary': 'A dummy package of cat',
           'version': '1.0'}],
 'links': {'first': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=100&offset=0&search=cat',
           'last': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=100&offset=0&search=cat'},
 'meta': {'count': 1, 'limit': 100, 'offset': 0}}

In [4]: app.content_sources.rest_client.snapshots_api.list_snapshot_rpms('b342c4f3-c0bd-4eb4-9b2d-aad494928d2d', limit=4, offset=2)
2024-03-13 18:35:03.202 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<>, params=[('limit', 4), ('offset', 2)]
Out[4]: 
{'data': [{'arch': 'noarch',
           'epoch': '0',
           'name': 'cat',
           'release': '1',
           'summary': 'A dummy package of cat',
           'version': '1.0'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'cheetah',
           'release': '5',
           'summary': 'A dummy package of cheetah',
           'version': '1.25.3'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'chimpanzee',
           'release': '1',
           'summary': 'A dummy package of chimpanzee',
           'version': '0.21'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'cockateel',
           'release': '1',
           'summary': 'A dummy package of cockateel',
           'version': '3.1'}],
 'links': {'first': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=4&offset=0',
           'last': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=4&offset=32',
           'next': '/api/content-sources/v1/snapshots/b342c4f3-c0bd-4eb4-9b2d-aad494928d2d/rpms?limit=4&offset=6'},
 'meta': {'count': 35, 'limit': 4, 'offset': 2}}

In [5]: 

`

Thank you

**Pull request merged**